### PR TITLE
add amqp case study

### PR DESCRIPTION
**Pull request by jmspring**

### CODE REVIEW

**Comment by sedouard** [comment]
fyi - I'm currently reviewing this case study @jmspring

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+The NodeJS module based upon the existing node-qpid module started as an extension to the existing module, first enabling asynchronous operations then extending it to allow higher throughput. This work exposed two challenges. First, due to the reliance on compiled code QPID Proton, building for more than Linux was a challenge. Second, adding in full asynchronous support exposed the lack of thread safety of the particular interface being used in QPID Proton. The updated node-qpid can be found <a href="https://github.com/jmspring/node-qpid">here</a>.
```
Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: You start talking about challenges here but there's a challenges section below. May be worth leaving things off with how it works and consolidate all the challenges in the #challenges section.

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Challenges
+
+As mentioned previously, the development of the node-qpid module relies on a build process that was tailored for Unix-based platforms, which made building on Windows a bit of a challenge. Since then, the QPID Proton library has been updated to work better with Windows. At this time, node-qpid has not yet integrated with these updates. Also, it has not yet had a pull request into the original module developed by Paul O’Fallon. Given the significant changes, the question remains to do a pull request or rename the module and create a new project.
```
Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: You should mention that Siemens needed the native module building for windows and maybe why.

**Comment by sedouard** [comment]
https://github.com/noodlefrenzy/node-sbus-amqp10/issues/3

**Comment by sedouard** [comment]
https://github.com/noodlefrenzy/node-sbus-amqp10/issues/3

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+## Opportunities for Reuse
+
+Any customer wanting to leverage Azure Service Bus and Event Hub from their NodeJS infrastructure can pick up these libraries and start working right away. In addition, anyone attempting to work with AMQP 1.0 messaging service, such as <a href="https://qpid.apache.org">qpidd</a>, can leverage the AMQP1.0 modules. Example are included within the modules themselves, for both <a href="https://github.com/noodlefrenzy/node-sbus-amqp10/blob/master/examples/simple_eventhub_client.js">Azure</a> and <a href="https://github.com/noodlefrenzy/node-amqp10/blob/master/examples/simple_activemq_test.js">non-Azure</a> scenarios.
```
Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: It woudln't hurt to embed some of the code sample here. For example I think its neat that you're 'plugging in' to the `sbus` module and returning a similar interface with new capability. From what I see, the code sample would be super tiny like:

Now its as easy as adding an npm module, which wraps the existing azure service bus module [`sbus`](https://npmjs.org/sbus) with amqp10 abilities:
```js
// get event hub client with amqp10 abilities
var client = require('sbus-amqp10').EventHubClient()
client.getEventProcessor(consumerGroup, function (conn_err, processor) {
  // initialize processor with callback fired for each message
  processor.init(function (rx_err, partition, payload) {
      console.log('received event payload:')
      console.dir(JSON.parse(payload);
  });
  // initiate listening to event hub via amqp10
  processor.recieve();
});
```

Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: I think you should also highlight that this works both on windows and linux. thats cool.

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md (hunk 2 above)
Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: `has been updated to work better with Windows` - Was this due to the work done here? It would be great if you snuck a link to a a PR  here

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+In order to develop a viable solution for Siemens, we started with two different approaches to implementing AMQP 1.0 for NodeJS - a pure NodeJS (NodeJS with no “native” modules) implementation and a higher performance implementation that started as modifications to the node-qpid module mentioned previously. The pure NodeJS module, <a href="https://www.npmjs.com/package/amqp10">node-amqp10</a> used the built-in Buffers within NodeJS as well as a few native-free modules (buffer-lists, node-int64, etc.) to implement the AMQP 1.0 binary protocol stack. It uses internal finite state-machines based on Stately.js, processing incoming payloads into connection/session/link connections/disconnects, and propagates results using fine-scale eventing. An outer “AMQP Client” class wraps all of the fine-grained details into a course-grained API (connect, send, receive), and that class is configured via a policy class allowing fine-grained control of almost all features, but with a deep inheritance/merge model allowing the baseline behaviors to be overridden as needed (see e.g. the EventHub policy vs. the baseline).
```
Comment (sedouard) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: Maybe a paragraph break at `An outer "AMQP Client...`. To me its a bit wordy here. Or consider trimming some unnecessary bits.

**Comment by sedouard** [comment]
gonna give this a shot this afternoon. It looks awesome though I think this is pretty valuable stuff and enabling using amqp eventhub from node is neat.

I'll follow up with anything I notice from trying it out and give you the magic phrase.

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+categories: AMQP NodeJS Azure "Event Hub" "Service Bus"
+color: "blue"
+#image: "{{ site.baseurl }}/images/imagename.png" #should be ~350px tall
+excerpt: Enabling Event Hub in NodeJS
+--- 
+## Overview
+
+This case study focuses on enabling the use of Service Bus and Event Hub from NodeJS based applications. Anyone interested in using Azure for cloud based queues, topics, or high scale event processing will benefit from the code and design discussed herein.
+
+In this case study, you will learn to leverage the libraries we have built to integrate Azure Service Bus and Event Hub into your NodeJS applications.
```
Comment (stephbu) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: Yeah this is the "attract mode" snippet that I think the GrapheneDB needs ;-)

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  
+## Customer Problem
+
+Siemens Energy was looking to deploy a fleet of NodeJS-based services onto Azure which would use Service Bus queues for inter-process communication and data marshaling, and Event Hub for log streaming for potential aggregation and alerting tasks. Current existing solutions for using Event Hub and Service Bus were tied to either Java or C#. Both Event Hub and Service Bus require AMQP 1.0 to enable the full range of functionality. A realistic AMQP 1.0 solution for NodeJS did not exist.
```
Comment (timfpark) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: Move the customer problem up to the top - this is the "why" that motivates the rest.

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#author-image: "{{ site.baseurl }}/images/Jim Spring/photo.jpg" //should be square dimensions
+date:   2015-08-26 17:39:15
+categories: AMQP NodeJS Azure "Event Hub" "Service Bus"
+color: "blue"
+#image: "{{ site.baseurl }}/images/imagename.png" #should be ~350px tall
+excerpt: Enabling Event Hub in NodeJS
+--- 
+## Overview
+
+This case study focuses on enabling the use of Service Bus and Event Hub from NodeJS based applications. Anyone interested in using Azure for cloud based queues, topics, or high scale event processing will benefit from the code and design discussed herein.
```
Comment (timfpark) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: Move this to the bottom and into a section on "how to get started" with the module.

File: _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
+---
+layout: post
+title:  "Enabling Event Hub in NodeJS"
```
Comment (timfpark) [inline _posts/2015-08-26-Enabling-EventHubs-In-NodeJS.md]: You might consider adding AMQP 1.0 into the title for SEO.  Perhaps "Building AMQP 1.0 node.js module for Azure Event Hub"?

**Comment by cicorias** [comment]
This looks like it's had 2 peer reviewers yet no looks good..
